includes/class-fs-garbage-collector.php

```diff
@@ -0,0 +1,359 @@
+
+        function run() {
+            require_once WP_FS__DIR_INCLUDES . '/class-fs-lock.php';
+
+            $lock = new FS_Lock( 'garbage_collection' );
+
+            if ( $lock->is_locked() ) {
+                return;
+            }
+
+            // Create a 1-day lock.
+            $lock->lock( WP_FS__TIME_24_HOURS_IN_SEC );
+
+            $this->clean_up();
+       }
+
+        private function clean_up() {
+            $_accounts = FS_Options::instance( WP_FS__ACCOUNTS_OPTION_NAME, true );
+
+            $products_to_delete = array();
+
+            foreach ( $this->get_product_types() as $product_type ) {
+                $products_to_delete[ $product_type ] = $this->get_products_to_delete( $_accounts, $product_type );
+            }
+
+            $update_options = false;
+
+            if ( ! empty( $products_to_delete ) ) {
+                if ( $this->delete_products( $_accounts, $products_to_delete ) ) {
+                    $this->delete_inactive_users( $_accounts );
+
+                    $update_options = true;
+                }
+            }
+
+            if ( $update_options ) {
+                $_accounts->store();
+            }
+        }
+
+        /**
+         * @param FS_Options $_accounts
+         * @param string     $product_type
+         *
+         * @return array
+         */
+        private function get_products_to_delete( $_accounts, $product_type ) {
+            $option_name      = ( $product_type . 's' );
+            $all_product_data = $_accounts->get_option( $option_name, array() );
+
+            $products_to_delete = array();
+
+            $has_updated_option = false;
+
+            foreach ( $all_product_data as $slug => $product_data ) {
+                if ( ! is_object( $product_data ) ) {
+                    continue;
+                }
+
+                if (
+                    empty( $product_data->last_load_timestamp ) ||
+                    ! is_numeric( $product_data->last_load_timestamp )
+                ) {
+                    // Set to the current time so that if the product is no longer used, its data will be deleted after 4 weeks.
+                    $product_data->last_load_timestamp = time();
+
+                    $has_updated_option = true;
+
+                    continue;
+                }
+
+                if ( $this->is_product_active( $slug, $product_data ) ) {
+                    continue;
+                }
+
+                $is_addon = FS_Plugin::is_valid_id( $product_data->parent_plugin_id );
+
+                if ( ! $is_addon ) {
+                    $products_to_delete[] = $product_data;
+                } else {
+                    /**
+                     * If add-on, add to the beginning of the array so that add-ons are removed before their parent. This is to prevent an unexpected issue when an add-on exists but its parent was already removed.
+                     */
+                    array_unshift( $products_to_delete, $product_data );

```

Comment:
@fajardoleo If a theme has add-ons, this logic might not hold and by your comment, it can still produce issues.

So I propose to completely separate this step. Instead of doing it here, you can do the rearranging inside the `delete_products` steps where it will iterate over all products and slice the add-ons and the rearrange by putting them all on top.